File: chapter_2/typescript/main/SelectionSort.ts
export default class SelectionSort {
  findLow(arr: number[]): number {
    const indexOfMin: number = arr.reduce((minIndex, current, currentIndex) => {
      return this.currentLessThanPrevious(current, arr[minIndex])
        ? currentIndex
        : minIndex;
    }, 0);
    return indexOfMin;
  }

  findHigh(arr: number[]): number {
    const indexOfMax: number = arr.reduce((minIndex, current, currentIndex) => {
      return this.currentGreaterThanPrevious(current, arr[minIndex])
        ? currentIndex
        : minIndex;
    }, 0);
    return indexOfMax;
  }

  selectionSort(arr: number[], order: number): number[] | number {
    if (this.isArrayEmpty(arr)) {
      return -1;
    }

    const sortedArr: number[] = [];

    switch (order) {
      case 1:
        while (arr.length > 0) {
          sortedArr.push(...arr.splice(this.findLow(arr), 1));
        }
        break;
    
      case -1:
        while (arr.length > 0) {
          sortedArr.push(...arr.splice(this.findHigh(arr), 1));
        }
        break;
    }

    return sortedArr;
  }

  isArrayEmpty(arr: number[]): boolean {
    if (arr.length === 0) {
      return true;
    }
    return false;
  }

  currentLessThanPrevious(current: number, previous: number): boolean {
    if (current < previous) {
      return true;
    }
    return false;
  }

  currentGreaterThanPrevious(current: number, previous: number): boolean {
    if (current > previous) {
      return true;
    }
    return false;
  }
}

File: chapter_2/golang/SelectionSort.go
package main

import "fmt"

func findLow(arr []int) int {
	low := arr[0]
	lowIndex := 0
	for i := 0; i < len(arr); i++ {
		if arr[i] < low {
			low = arr[i]
			lowIndex = i
		}
	}
	return lowIndex
}

func selectionSort(arr []int) []int {
	size := len(arr)
	sortedArr := make([]int, size)
	for i := 0; i < size; i++ {
		smallest := findLow(arr)
		sortedArr[i] = arr[smallest]
		arr = append(arr[:smallest], arr[smallest+1:]...)
	}
	return sortedArr
}

func main() {
	fmt.Print(selectionSort([]int{7, 8, 5, 32, 4, 9}))
}
